webapp/src/Service/ScoreboardService.php

```diff
@@ -499,6 +519,31 @@ public function calculateScoreRow(
         if ($updateRankCache && ($correctJury || $correctPubl)) {
             $this->updateRankCache($contest, $team);
         }
+
+        // If we did not have a first to solve, we need to check if we have any
+        // potential first to solve that are now the first to solve.
+        // We only do this if there are no pending submissions for this problem
+        // for this team and if this submission is not a potential first to solve itself.
+        // We also do this only once, to not have infinite loops if we have multiple
+        // potential first to solves.
+        if ($updatePotentialFirstToSolves && $pendingJury === 0 && !$potentialFirstToSolve) {
+            /** @var ScoreCache[] $potentialFirstToSolves */
+            $potentialFirstToSolves = $this->em->createQueryBuilder()
+                ->from(ScoreCache::class, 's')
+                ->join('s.team', 't')
+                ->select('s', 't')
+                ->andWhere('s.contest = :contest')
+                ->andWhere('s.problem = :problem')
+                ->andWhere('s.is_potential_first_to_solve = 1')
```

Comment:
Is this restriction actually adding anything? We are hopefully only very rarely executing this codepath anyway, so it would not hurt to go here over all teams
- which have solved this problem (which is a restriction you are currently not adding)
- after us (right?)

Reply:
We only set `s. is_potential_first_to_solve` if that team solved the problem and we have earlier pending submissions, so I'd say that we do filter on teams solving this problem.

My current code assumes that if we loop over all potentional first to solves for the same problem, we will find the one that we need to actually update.

Currently this code path is executed more than you think: it is executed when we have no pending jury submissions for this (team,problem) and we are not the first to solve ourselves.

If you have a better idea to implement this, I am open, but I don't follow your thought currently. Feel free to push it as a commit. I think the added test should catch if it works.

Reply:
I am thinking of something like https://github.com/DOMjudge/domjudge/commit/76f1640d4edc3d30adcbfe86911cd54acc4bb7d6

That should make sure we don't call it too often while not adding the complexity of the extra column. I still have to think through the potential race conditions...

Reply:
Something like that could indeed work I think.